account-kit/react/src/hooks/useSolanaTransaction.ts

```diff
@@ -97,34 +118,84 @@ export function useSolanaTransaction(
-      const transaction = policyId
-        ? await signer.addSponsorship(instructions, connection, policyId)
-        : await signer.createTransfer(instructions, connection);
+    mutationFn: async ({
+      transactionComponents: {
+        preSend,
+        transformInstruction = mapTransformInstructions.default,
+      } = {},
+      ...params
+    }: SolanaTransactionParams) => {
+      const instructions = getInstructions();
+      let transaction: VersionedTransaction | Transaction =
+        await transformInstruction(instructions);
 
-      await signer.addSignature(transaction);
+      transaction = (await preSend?.(transaction)) || transaction;
 
-      const hash = await solanaNetwork.broadcast(connection, transaction);
+      if (needsSignerToSign()) {
+        await signer?.addSignature(transaction);
+      }
+
+      const localConnection = connection || missing("connection");
+      const hash = await solanaNetwork.broadcast(localConnection, transaction);
       return { hash };
+
+      function getInstructions() {
+        if ("instructions" in params) {
+          return params.instructions;
+        }
+        return [
+          SystemProgram.transfer({
+            fromPubkey: new PublicKey(
+              signer?.address || missing("signer.address")
+            ),
+            toPubkey: new PublicKey(params.transfer.toAddress),
+            lamports: params.transfer.amount,
+          }),
+        ];
+      }
+
+      function needsSignerToSign() {
+        if ("message" in transaction) {
+          const message = transaction.message;
+          return message.staticAccountKeys.some(
+            (key, index) =>
+              key.toBase58() === signer?.address &&
+              message?.isAccountSigner(index)
+          );
+        }
+        return transaction.instructions.some((x) =>
+          x.keys.some(
+            (x) => x.pubkey.toBase58() === signer?.address && x.isSigner
+          )
+        );
+      }
     },
     ...opts.mutation,
   });
   const signer: null | SolanaSigner = opts?.signer || fallbackSigner;
   const connection = opts?.connection || backupConnection?.connection || null;
+  const policyId =
+    "policyId" in opts ? opts.policyId : backupConnection?.policyId;
+  const mapTransformInstructions: Record<string, TransformInstruction> = {
+    async addSponsorship(instructions: TransactionInstruction[]) {
+      return await (signer || missing("signer")).addSponsorship(
+        instructions,
+        connection || missing("connection"),
+        policyId || missing("policyId")
+      );
+    },
+    async createTransfer(instructions: TransactionInstruction[]) {
+      return await (signer || missing("signer")).createTransfer(
+        instructions,
+        connection || missing("connection")
+      );
+    },
+    get default() {
+      return policyId
+        ? mapTransformInstructions.addSponsorship
+        : mapTransformInstructions.createTransfer;
```

Comment:
oh just saw this, what's `createTransfer` as default here? should this just be createTransaction?

curious on the flow above too, should we not just take `instructions`, put them inside of a tx, add sponsorhip optionally, apply preSend, apply our sig, then submit?

maybe that's what's already happening with `createTransfer` and the naming threw me off

Reply:
Naming threw you off, should I change